builtin/logical/transit/path_key_export.go

```diff
@@ -0,0 +1,186 @@
+				Description: "Name of the key",
+			},
+		},
+
+		Callbacks: map[logical.Operation]framework.OperationFunc{
+			logical.ReadOperation: b.pathPolicyExport,
+		},
+
+		HelpSynopsis:    pathExportHelpSyn,
+		HelpDescription: pathExportHelpDesc,
+	}
+}
+
+func (b *backend) pathPolicyExport(
+	req *logical.Request, d *framework.FieldData) (*logical.Response, error) {
+	exportType := d.Get("export_type").(string)
+	name := d.Get("name").(string)
+	version := d.Get("version").(string)
+
+	switch exportType {
+	case exportTypeEncryptionKey:
+	case exportTypeSigningKey:
+	case exportTypeHMACKey:
+	default:
+		return logical.ErrorResponse(fmt.Sprintf("invalid export type: %s", exportType)), logical.ErrInvalidRequest
+	}
+
+	p, lock, err := b.lm.GetPolicyShared(req.Storage, name)
+	if lock != nil {
+		defer lock.RUnlock()
+	}
+	if err != nil {
+		return nil, err
+	}
+	if p == nil {
+		return nil, nil
+	}
+
+	if !p.Exportable {
+		return logical.ErrorResponse("key is not exportable"), nil
+	}
+
+	switch exportType {
+	case exportTypeEncryptionKey:
+		if !p.Type.EncryptionSupported() {
+			return logical.ErrorResponse("encryption not supported for the key"), logical.ErrInvalidRequest
+		}
+	case exportTypeSigningKey:
+		if !p.Type.SigningSupported() {
+			return logical.ErrorResponse("signing not supported for the key"), logical.ErrInvalidRequest
+		}
+	}
+
+	resp := &logical.Response{
+		Data: map[string]interface{}{
+			"name": p.Name,
+		},
+	}
+
+	if version != "" {
+		var versionValue int
+		if version == "latest" {
+			versionValue = p.LatestVersion
+		} else {
+			version = strings.TrimPrefix(version, "v")
+			versionValue, err = strconv.Atoi(version)
+			if err != nil {
+				return logical.ErrorResponse("invalid key version"), logical.ErrInvalidRequest
+			}
+		}
+		resp.Data["version"] = versionValue
+
+		key, ok := p.Keys[versionValue]
+		if !ok {
+			return logical.ErrorResponse("version does not exist or is no longer valid"), logical.ErrInvalidRequest
+		}
+
+		exportKey, err := getExportKey(p, key, exportType)
+		if err != nil {
+			return nil, err
+		}
+		resp.Data["key"] = exportKey
+
+		return resp, nil
+	}
+
+	retKeys := map[string]string{}
+	for k, v := range p.Keys {
+		exportKey, err := getExportKey(p, v, exportType)
+		if err != nil {
+			return nil, err
+		}
+		retKeys[strconv.Itoa(k)] = exportKey
+	}
+	resp.Data["keys"] = retKeys
+
+	return resp, nil
+}
+
+func getExportKey(policy *keysutil.Policy, key keysutil.KeyEntry, exportType string) (string, error) {
+	switch exportType {
+	case exportTypeHMACKey:
+		return strings.TrimSpace(base64.StdEncoding.EncodeToString(key.HMACKey)), nil
+	case exportTypeEncryptionKey:
+		switch policy.Type {
+		case keysutil.KeyType_AES256_GCM96:
+			return strings.TrimSpace(base64.StdEncoding.EncodeToString(key.AESKey)), nil
+		}
+	case exportTypeSigningKey:
+		switch policy.Type {
+		case keysutil.KeyType_ECDSA_P256:
+			ecKey, err := keyEntryToECPrivateKey(key)
```

Comment:
I think it'd be good to pass the curve in here to keep the keyEntry function generic, so if we support P521 for instance you can just pass the curve in there rather than hardcode the curve in the function itself.